pulsar/producer_partition.go

```diff
@@ -256,10 +256,17 @@ func (p *partitionProducer) internalSend(request *sendRequest) {
 	}
 
 	if sendAsBatch {
-		ok := p.batchBuilder.Add(smm, sequenceID, msg.Payload, request, msg.ReplicationClusters)
-		if ok == false {
+		added := p.batchBuilder.Add(smm, sequenceID, msg.Payload, request, msg.ReplicationClusters)
+		if !added {
 			// The current batch is full.. flush it and retry
 			p.internalFlushCurrentBatch()
+
+			// after flushing try again to add the current payload
+			if ok := p.batchBuilder.Add(smm, sequenceID, msg.Payload, request, msg.ReplicationClusters); !ok {
+				p.log.WithField("size", len(msg.Payload)).
+					WithField("sequenceID", sequenceID).
+					Error("unable to send message")
+			}
 		}
 	} else {
 		// Send individually
```

Comment:
I think there may be the same problem here. If false is returned after the https://github.com/apache/pulsar-client-go/blob/master/pulsar/internal/batch_builder.go#L107 function is executed.

```
// Add will add single message to batch.
func (bb *BatchBuilder) Add(metadata *pb.SingleMessageMetadata, sequenceID uint64, payload []byte,
	callback interface{}, replicateTo []string) bool {
	if replicateTo != nil && bb.numMessages != 0 {
		// If the current batch is not empty and we're trying to set the replication clusters,
		// then we need to force the current batch to flush and send the message individually
		return false
	} else if bb.msgMetadata.ReplicateTo != nil {
		// There's already a message with cluster replication list. need to flush before next
		// message can be sent
		return false
	} else if bb.hasSpace(payload) {
		// The current batch is full. Producer has to call Flush() to
		return false
	}

	if bb.numMessages == 0 {
		bb.msgMetadata.SequenceId = proto.Uint64(sequenceID)
		bb.msgMetadata.PublishTime = proto.Uint64(TimestampMillis(time.Now()))
		bb.msgMetadata.SequenceId = proto.Uint64(sequenceID)
		bb.msgMetadata.ProducerName = &bb.producerName
		bb.msgMetadata.ReplicateTo = replicateTo

		bb.cmdSend.Send.SequenceId = proto.Uint64(sequenceID)
	}
	addSingleMessageToBatch(bb.buffer, metadata, payload)

	bb.numMessages++
	bb.callbacks = append(bb.callbacks, callback)
	return true
}
```

 This data will not be retried, resulting in incorrect processing. Therefore, I think retry should be added after internalFlushCurrentBatch under this condition.

Reply:
You think we need to retry in the send individually block as well?

Reply:
Yes, I think it is ok to try again, at least a warning should be issued to inform the user that a message cannot be correctly added to the buffer.

Reply:
Agree with @tuteng 

```
if add := p.batchBuilder.Add(smm, sequenceID, msg.Payload, request, msg.ReplicationClusters); !add {
    //print warn log
}
```

Reply:
Agree with @tuteng , we need to handle the return value of `p.batchBuilder.Add()`:

```
if add := p.batchBuilder.Add(smm, sequenceID, msg.Payload, request, msg.ReplicationClusters); !add {
    // print warn log
}
```

Reply:
Sounds good. I'll add a log there.